report/renderer/report-utils.js

```diff
@@ -187,6 +190,75 @@ class ReportUtils {
     return clone;
   }
 
+  /**
+   * Given an audit's details, identify and return a URL locator function that
+   * can be called later with an `item` to extract the URL of it.
+   * @param {LH.FormattedIcu<LH.Audit.Details.TableColumnHeading[]>} headings
+   * @return {{(item: LH.FormattedIcu<LH.Audit.Details.TableItem>): string|undefined}=}
+   */
+  static getUrlLocatorFn(headings) {
+    // The most common type, valueType=url.
+    const urlKey = headings.find(heading => heading.valueType === 'url')?.key;
+    if (urlKey && typeof urlKey === 'string') {
+      // Return a function that extracts item.url.
+      return (item) => {
+        const url = item[urlKey];
+        if (typeof url === 'string') return url;
+      };
+    }
+
+    // The second common type, valueType=source-location.
+    const srcLocationKey = headings.find(heading => heading.valueType === 'source-location')?.key;
+    if (srcLocationKey) {
+      // Return a function that extracts item.source.url.
+      return (item) => {
+        const sourceLocation = item[srcLocationKey];
+        if (typeof sourceLocation === 'object' && sourceLocation.type === 'source-location') {
+          return sourceLocation.url;
+        }
+      };
+    }
+
+    // More specific tests go here, as we need to identify URLs in more audits.
+  }
+
+  /**
+   * Mark TableItems/OpportunityItems with entity names.
+   * @param {LH.Result.Entities|undefined} entityClassification
+   * @param {import('../../types/lhr/audit-result').Result} audit
+   */
+  static classifyEntities(entityClassification, audit) {
+    if (!entityClassification) return;
+    if (audit.details?.type !== 'opportunity' && audit.details?.type !== 'table') {
```

Comment:
Intent might be clearer if these checks are moved out into `prepareReportResult`, so as you're reading that method you know you only need to step into `classifyEntities` if it's a table or opportunity (right now you have to trust the comment isn't out of date).

In here, the params could then be narrowed, dropping the `undefined` on `entityClassification`, and audit could become details of `'LH.Audit.Details.Opportunity'|'LH.Audit.Details.Table'`

Reply:
These are to keep TS happy in the rest of the function, as we're passing the full audit here. Is there a better way to do it?

Reply:
I mean something like
```diff
diff --git a/report/renderer/report-utils.js b/report/renderer/report-utils.js
index a38b52d2b..739cc6210 100644
--- a/report/renderer/report-utils.js
+++ b/report/renderer/report-utils.js
@@ -91,7 +91,11 @@ class ReportUtils {
         }
 
         // Attach table/opportunity items with entity information.
-        ReportUtils.classifyEntities(result.entities, audit);
+        if (result.entities && audit.details) {
+          if (audit.details.type === 'opportunity' || audit.details.type === 'table') {
+            ReportUtils.classifyEntities(result.entities, audit.details);
+          }
+        }
 
         // TODO: convert printf-style displayValue.
         // Added:   #5099, v3
@@ -224,17 +228,12 @@ class ReportUtils {
 
   /**
    * Mark TableItems/OpportunityItems with entity names.
-   * @param {LH.Result.Entities|undefined} entityClassification
-   * @param {import('../../types/lhr/audit-result').Result} audit
+   * @param {LH.Result.Entities} entityClassification
+   * @param {LH.FormattedIcu<LH.Audit.Details.Opportunity|LH.Audit.Details.Table>} details
    */
-  static classifyEntities(entityClassification, audit) {
-    if (!entityClassification) return;
-    if (audit.details?.type !== 'opportunity' && audit.details?.type !== 'table') {
-      return;
-    }
-
+  static classifyEntities(entityClassification, details) {
     // If details.items are already marked with entity attribute during an audit, nothing to do here.
-    const {items, headings} = audit.details;
+    const {items, headings} = details;
     if (!items.length || items.some(item => item.entity)) return;
 
     // Identify a URL-locator function that we could call against each item to get its URL.
```

Reply:
So I was passing `audit.details` or more specific params earlier, I think. But I now think it would be beneficial to pass the full `audit` to get the `id` as well, in case if any audit needs special handling in future.